calyx/src/analysis/static_par_timing.rs

```diff
@@ -0,0 +1,321 @@
+        writeln!(
+            f,
+            "This maps ids of par blocks to \" cell timing maps \", which map cells to intervals (i,j), that signify the clock cycles the group is active for, \n relative to the start of the given par block"
+        )?;
+        write!(
+            f,
+            "============ Map for Component \"{}\"",
+            self.component_name
+        )?;
+        writeln!(f, " ============")?;
+        let map = self.cell_map.clone();
+        // Sorting map to get deterministic ordering
+        let mut vec: Vec<(u64, ThreadTimingMap)> = map.into_iter().collect();
+        vec.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));
+        for (par_id, thread_timing_map) in vec.into_iter() {
+            write!(f, "========")?;
+            write!(f, "Par Node ID: {:?}", par_id)?;
+            writeln!(f, " ========")?;
+            let mut vec1: Vec<(u64, CellTimingMap)> =
+                thread_timing_map.into_iter().collect::<Vec<_>>();
+            vec1.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));
+            for (thread_id, cell_timing_map) in vec1 {
+                write!(f, "====")?;
+                write!(f, "Child/Thread ID: {:?}", thread_id)?;
+                writeln!(f, " ====")?;
+                let mut vec2: Vec<(ir::Id, Vec<(u64, u64)>)> =
+                    cell_timing_map.into_iter().collect::<Vec<_>>();
+                vec2.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));
+                for (cell_name, clock_intervals) in vec2 {
+                    write!(f, "{:?} -- ", cell_name)?;
+                    writeln!(f, "{:?}", clock_intervals)?;
+                }
+            }
+            writeln!(f)?
+        }
+        write!(f, "")
+    }
+}
+
+impl StaticParTiming {
+    /// Construct a live range analysis.
+    pub fn new(
+        control: &mut ir::Control,
+        comp_name: ir::Id,
+        live: &LiveRangeAnalysis,
+    ) -> Self {
+        let mut time_map = StaticParTiming {
+            component_name: comp_name,
+            ..Default::default()
+        };
+
+        // compute_unique_ids is deterministic
+        // so if we're calling it after we've called live range analysis (and the
+        // control program hasn't changed, then this is unnecessary)
+        ControlId::compute_unique_ids(control, 0, false);
+
+        time_map.build_time_map(control, None, live);
+
+        time_map
+    }
+
+    /// `par_id` is the id of a par thread.
+    /// `thread_a` and `thread_b` are ids of direct children of par_id (if `thread_a` and
+    /// `thread_b` are *not* direct children of par_id, then the function will error)
+    /// `a` and `b` are cell names
+    /// liveness_overlaps checks if the liveness of `a` in `thread_a` ever overlaps
+    /// with the liveness of `b` in `thread_b`
+    /// if `par_id` is not static, then will automtically return true
+    pub fn liveness_overlaps(
+        &self,
+        par_id: &u64,
```

Comment:
One common pattern used when building such interfaces is the [NewType Idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). Essentially, if your interface produces these `u64` tokens and consumes them, you can construct a new struct that wraps the `u64` and doesn't allow the user code to look into it. That way, the *only* thing user code can do is get these node and thread ID tokens from the analysis and send them back to perform various queries. 

The `ir/id.rs` shows how we use this for representing identifier using simple wrapped numbers.

Reply:
I think I understand what you're saying: we should have a struct `Par(u64)` and `ParThread(u64)`, and the `cell_share` pass should never actually look at the u64 that's inside these structs. 

One thing I should probably mention: in the cell-share pass, the `Par(u64)` and `ParThread(u64)` tokens would actually be generated from the LiveRangeAnalysis Analysis, whereas the `Par(u64)` and `ParThread(u64)` tokens we deal with here are obviously just generated by the ParStaticTiming Analysis. 
Currently,  I'm relying on the fact that `ControlId` generates a deterministic labeling of node IDs, so the labeling of IDs will be the same for both analyses. Do you think it would be more "solid" to change `ControlId` so that we can actually instantiate it, and so we could literally pass the same `ControlId` instance through both the Live Range Analysis and StaticParTiming Analysis? This wouldn't really affect correctness, but it does seem more "solid".

Reply:
Yeah, I think having `ControlId` actually provide those "wrapped tokens" is definitely a way better interface. An even funkier trick that people use in rust is to actually associate all produced tokens with the analysis by giving it a fake lifetime:
```
pub struct WrappedIdx<'a> {
  idx: u64,
  _d: std::marker::PhantomData<'a>
}

pub struct LiveRanges<'a> {
  pub fn get_label(node: &ir::Id) -> WrappedIdx<'a> ...
}
```